Allow pipeline writer to spit out Ogg directly, including when seeking

philippe44 opened the pull request:
Use ogg v 0.8 to use directly inner writer instead of the complicated workaround in previous patch https://github.com/librespot-org/librespot/pull/560

sashahilton00 commented:
I've merged a fix for the CI, and updated the Cargo.lock file, along with bumping the minimum rust version to `1.46.0`, since the current rust version is `1.49.0`, and `vec_drain_as_slice` was introduced in `1.46.0` and thus was causing issues during compilation.

philippe44 commented:
Thanks!

ashthespy commented:
> along with bumping the minimum rust version to `1.46.0`, since the current rust version is `1.49.0`, and `vec_drain_as_slice` was introduced in `1.46.0` and thus was causing issues during compilation.

Shouldn't this then be a feature flag gated feature? 
For the core library, it would be nice to maintain Debian's current stable (buster) version of Rust as the MSRV (1.41).

sashahilton00 commented:
@ashthespy probably not a bad idea. I personally don't care about keeping the supported compiler version low, but I can see the argument for it. @philippe44 are you ok to add in a compile time flag for this feature and drop the CI MSRV back to `1.41.0`?

philippe44 commented:
I'll have to look at how to make that a clean compile time option (I'm a Rust noob, so everything requires learning) or I'll see if I can use another way to get the data instead of using vec_drain_as_slice

plietar commented:
Isn’t that code just equivalent to `vec![data[0] as i16]`?

plietar commented:
I kind of worry about passing the OGG data as i16s. For instance in this particular line, an extraneous byte gets inserted in the u8 -> i16 -> u8 conversion process.

I think it would be more appropriate to turn AudioPacket into a enum with a Vec<i16> entry for PCM samples and a Vec<u8> for OGG data.

The pipe backend matches on the packet type and does the right thing. Other backends crash on OGG data.

philippe44 commented:
> Isn’t that code just equivalent to `vec![data[0] as i16]`?
>
I'm still a Rust noob and I had a hard time with the first PR that forced me to use Ref/RefCell/Arc/interior mutation :-(

philippe44 commented:
> I kind of worry about passing the OGG data as i16s. For instance in this particular line, an extraneous byte gets inserted in the u8 -> i16 -> u8 conversion process.
> 
> I think it would be more appropriate to turn AudioPacket into a enum with a Vec entry for PCM samples and a Vec for OGG data.
> 
I agree but this is always the last byte of a stream so it will be dropped. I did that b/c I did not want to have even more change to the rest of the code. I can study that option if you prefer
> The pipe backend matches on the packet type and does the right thing. Other backends crash on OGG data.

philippe44 commented:
Looking at the pipe.rs code, it seems that it expects Vec<i16>
```
impl Sink for StdoutSink {
    fn start(&mut self) -> io::Result<()> {
        Ok(())
    }

    fn stop(&mut self) -> io::Result<()> {
        Ok(())
    }

    fn write(&mut self, data: &[i16]) -> io::Result<()> {
        let data: &[u8] = unsafe {
            slice::from_raw_parts(
                data.as_ptr() as *const u8,
                data.len() * mem::size_of::<i16>(),
            )
        };

        self.0.write_all(data)?;
        self.0.flush()?;

        Ok(())
    }
}
```
I understand I can modify that as well, but how much are the maintainers willing to let such modification creep too far in the code?

sashahilton00 commented:
I was just reading up on rust enums, they are pretty elegant. However, in the interests of expediency do we want to merge this and evaluate the audio packet changes in a separate PR - such a change looks like it will require some rewriting of the existing back ends.

philippe44 commented:
I think I have been able to use enum properly, well at least it works, produce the expected result and does not have too many ripple effects. The only one is the addition of a write_bytes() trait that has a default version that does nothing.

I've done per asked, a samples() function to access i16 and a bytes() function to access raw data. The passthrough is now properly writing u8 and not adding (in some case) an extra byte.

As said before, excuse Rust rookie style & practices mistakes. I also have an issue with the merger that I'm not sure I now how to fix. My cargo.lock should never have been pushed and the audio/Cargo.toml should just add ogg, I don't know why it is in conflict now

philippe44 commented:
I don't think I can fix the --locked issue

philippe44 commented:
Any love/hate/opinion for that version?

sashahilton00 commented:
I'll try and fix the cargo.lock file tomorrow. There are a few PRs which are having problems due to older rust versions, to the point where it may make sense to just bump the MSRV to `1.42.0`. With regards to the enums, @plietar is the enum implementation what you were expecting? Also, given the change to AudioPacket which is used in various backends, have they been tested to ensure that they are functioning as expected?

plietar commented:
> With regards to the enums, @plietar is the enum implementation what you were expecting?

Generally looks good. I'll have a proper look through the code this weekend.

sashahilton00 commented:
We should also update the instructions/dependencies in `COMPILING.md` wrt the Jack backend now using Rodio

sashahilton00 commented:
Are we good to merge this?

plietar reviewed (commented):
Could you add a high level description of how the passthru decoder works please.
In particular what is the purpose of the absgp stitching, and how does seeking work.

On every seek we re-send the same headers? Is that something decoders tolerate / require?

Johannesd3 commented on audio/src/lib.rs:
```diff
@@ -29,8 +30,65 @@ pub use fetch::{
+
+impl AudioPacket {
+    pub fn samples(&self) -> &[i16] {
+        match &*self {
```
If you go with pattern matching instead of using this functions (as @plietar suggested below), you could also remove this functions entirely. At least I wouldn't expect such harmless-looking functions to panic.

philippe44 replied:
@Johannesd3 - understood but back to my comment, then I would need to modify all backends to do the matching and I was trying to avoid that like plague

plietar replied:
@Johannesd3 I actually think these functions are useful inside the backends that only support one kind of packet. Of course each backend could also match and do the panic themselves, but it's cleaner in a version.

philippe44 replied:
Done

philippe44 commented on audio/src/lib.rs:
```diff
@@ -29,8 +30,65 @@ pub use fetch::{
+        }
+    }
+
+    pub fn len(&self) -> usize {
```
It's to simplify the usage in playback/src/player.rs where it's convenient to access len prior knowing the type of data underneath (handle_packet)

plietar replied:
Hmm, would an `empty(&self) -> bool` be enough in that case?

philippe44 replied:
you're torturing me but ok :-)

Johannesd3 commented on src/main.rs:
```diff
@@ -204,6 +204,11 @@ fn setup(args: &[String]) -> Setup {
             "",
             "disable-gapless",
             "disable gapless playback.",
+        )
+	    .optflag(
+            "",
+            "passthrough",
+            "Pass raw stream to output, only works for \"pipe\"."
```
I might overlook something, but shouldn't we check that the backend is indeed pipe if this option is present?

philippe44 replied:
Or, we could even force the backend to be pipe - that was my initial opinion

plietar replied:
I prefer an error than forcing the backend. It would be less surprising to users.

philippe44 replied:
ok

philippe44 commented on playback/src/audio_backend/mod.rs:
```diff
@@ -8,6 +8,9 @@ pub trait Sink {
     fn write(&mut self, data: &[i16]) -> io::Result<()>;
+    fn write_bytes(&mut self, _data: &[u8]) -> io::Result<()> {
+        Ok(())
+    }
```
I agree but then I would have to modify every existing backend which I would assume was not desirable, that's the reason I ended up with this. I did not want that PR to be a virus spreading across code.

Johannesd3 replied:
Would it be an option to add an optional type parameter to `Sink`? Like
```rust
pub trait Sink<S = i16> {
    fn write(&mut self, data: &[S]) -> io::Result<()>; 
}
```

plietar replied:
> I would have to modify every existing backend which I would assume was not desirable

Modifying every backend is perfectly fine if the overall result is nicer.

-----

@Johannesd3 I'm not sure this actually works. The player holds a `Box<dyn Sink>`, so it is important that all sinks expose the same interface.

Johannesd3 replied:
It works for certain, but not without modifying very much existing code, so probably it was a bad idea.

philippe44 replied:
> > I would have to modify every existing backend which I would assume was not desirable
> 
> Modifying every backend is perfectly fine if the overall result is nicer.
> 
So... are you really sure you want me to do that?

sashahilton00 replied:
I think it would be cleaner. Matching from each backend doesn't really add any code complexity, and it makes more sense than multiple methods to read/write the same underlying data, albeit in a different format.

philippe44 replied:
done

Johannesd3 commented on playback/src/player.rs:
```diff
@@ -378,7 +378,7 @@ enum PlayerPreload {
 }
 
-type Decoder = VorbisDecoder<Subfile<AudioDecrypt<AudioFile>>>;
+type Decoder = Box<dyn AudioDecoder + Send>;
```
There were already previously two different `VorbisDecoder`s, and it was solved with feature flags. Now, a third decoder is added, and a trait is used instead. It would be more consistent IMHO to use either feature flags or traits for all decoders. A third option (with less overhead than boxed traits) would be an `enum`, which is sometimes an alternative to traits if there's a well-known number of implementors.

plietar replied:
Feature flags don't work because we need to switch between the two based on runtime flags.
I don't really see a usecase for switching between libogg and lewton at runtime, so feature flags is fine there.

I agree there is a bit of inconsistency, but it seems reasonable and actually simpler this way.

I think the overhead of boxed trait is completely fine in this context.

philippe44 commented on audio/src/passthrough_decoder.rs:
```diff
@@ -0,0 +1,233 @@
+            if len > 0 {
+                let result = AudioPacket::Bytes(data.to_vec());
+                data.clear();
+                data.shrink_to_fit();
```
will do

philippe44 commented on audio/src/passthrough_decoder.rs:
```diff
@@ -0,0 +1,233 @@
+            let data = self.wtr.inner_mut();
+            let len = data.len();
+
+            if len > 0 {
```
It should not, it was just a sanity check but that can be avoided. I'll remove it

Johannesd3 replied:
It doesn't seem removed.

philippe44 replied:
oh this one, no it's on purpose. You might have nothing produced by the packet writer

philippe44 commented on audio/src/passthrough_decoder.rs:
```diff
@@ -0,0 +1,233 @@
+    return Ok(());
+}
+
+static mut BASEGP_PAGE: u64 = 0;
```
It is a static because it survives decoder instances and it is stitching together the global timeline. If not, multiple tracks will have an incorrect position and the resulting file will not be readable

plietar replied:
Ah right. So is the idea that for the lifetime of the process the granule only ever gets incremented?

Johannesd3 replied:
I don't understand what apsgp does, so I might be wrong. But librespot could be used as library, and multiple PassthroughDecoders could be created at the same time for whatever reason. What would happen then? There isn't any protection against concurrent access to `BASEGP_PAGE`, is it?

philippe44 replied:
> Ah right. So is the idea that for the lifetime of the process the granule only ever gets bigger?
>
Yes

philippe44 replied:
> I don't understand what apsgp does, so I might be wrong. But librespot could be used as library, and multiple PassthroughDecoders could be created at the same time for whatever reason. What would happen then? There isn't any protection against concurrent access to `BASEGP_PAGE`, is it?
>
There is probably no perfect answer. This relates to granule position's page. It is used to seek into file as well as a way to know the duration of an ogg file:  the granule of the last page of the file give you the duration, there is no other way. So if I don't have that incremented properly, then duration is wrong and players can be totally confused. There is no perfect solution anyway b/c when a seek is performed, the granule will overshoot, but that's better than undershoot that lead to premature ending. Re having that as a global, I agree and I don't like globals, but otherwise I would need to offer a "context" which is much more complications and non-compatibilty with existing code. I could have created another new() with an extra parameter to reset it but I think Rust does not offer overloading/mangling

Johannesd3 replied:
Ah, it's clear for me now.

But using `unsafe` in a library such that undefined behaviour could occur if you use it wrong should definitely be avoided.

You could use 
```rust

static BASEGP_PAGE: AtomicU64::new(0);

// Adding
BASEGP_PAGE.fetch_add(n, Ordering::Relaxed);
```

Even though it doesn't make sense to use multiple instance at the same time, it's at least no UB.

philippe44 replied:
done

philippe44 commented on audio/src/passthrough_decoder.rs:
```diff
@@ -0,0 +1,233 @@
+use super::{AudioDecoder, AudioError, AudioPacket};
+use ogg::{OggReadError, Packet, PacketReader, PacketWriteEndInfo, PacketWriter};
+use std::fmt;
+use std::io::{Error, Read, Seek};
```
done

philippe44 commented on audio/src/lib.rs:
```diff
@@ -29,8 +30,65 @@ pub use fetch::{
+
+pub enum AudioPacket {
+    Samples(Vec<i16>),
+    Bytes(Vec<u8>),
```
I was thinking that would more describe a blob so that if one day there is another codec, we could still refer to the outcome as 'bytes'

sashahilton00 replied:
I think `OggData` would be more self explanatory. If another codec is implemented at a later date, then we can think about renaming this, but for now Spotify encodes everything with `ogg` and doesn't look to be changing anytime soon.

philippe44 replied:
Done

philippe44 commented on audio/src/libvorbis_decoder.rs:
```diff
@@ -77,3 +72,9 @@ impl error::Error for VorbisError {
     }
 }
+
+impl From<vorbis::VorbisError> for AudioError {
```
Here I'm struggling ... I thought it would be define in libvorbis because it does that for AudioDecoder. And I can't find a proper option if I define it lib.rs (help!)

sashahilton00 replied:
Unless I'm missing something, it should be possible to define

```
#[cfg(feature = "with-tremor")]
extern crate librespot_tremor as vorbis;
#[cfg(not(feature = "with-tremor"))]
extern crate vorbis;
```
at the top of `lib.rs`, and then further down by `AudioError`, convert from the `VorbisError` using a `From` implementation:

```
impl From<vorbis::VorbisError> for AudioError {
    fn from(err: vorbis::VorbisError) -> AudioError {
        AudioError::VorbisError(VorbisError::from(err))
    }
}
```
And then remove the existing implementations from the `libvorbis`/`lewton` decoders

philippe44 commented on playback/src/player.rs:
```diff
@@ -1061,21 +1071,29 @@ impl PlayerInternal {
+                if packet.len() > 0 {
+                    let res;
+
+                    if !self.config.passthrough {
```
ok, but I think I need something like
```
	if let AudioPacket::Samples(ref mut data) = packet {
						if let Some(ref editor) = self.audio_filter {
							editor.modify_stream(data)
						}
 
						if self.config.normalisation && normalisation_factor != 1.0 {
							for x in data.iter_mut() {
								*x = (*x as f32 * normalisation_factor) as i16;
							}
						}
```
but I'm not sure, Rust is still giving me headaches at this point

philippe44 commented:
> Could you add a high level description of how the passthru decoder works please.
> In particular what is the purpose of the absgp stitching, and how does seeking work.
> 
> On every seek we re-send the same headers? Is that something decoders tolerate / require?
>
I had to review things properly after that discussion. I was struggling too much with Rust, using Ogg backend at the same time. So there was a fair bit of remaining issues that are hopefully corrected now in addition of complying to most of your requirements.

A a general comment, this patch is to help the (spotty)[https://github.com/michaelherger/spotty]  project where the main objective is to stream *only* one track per instance (with and optional starting offset) then launch another instance for the next track.

The passthru decoder nomally simply copies the entire stream to stdout, untouched. Now, there are a few problems because OggVorbis is not a fully streamable format as there are 3 headers (ident, comment and setup) per stream and they are all required especially the setup contains information to configure the vorbis decoder. Streams are divided into pages that have an own sequence number. There is no global header that tells the duration, instead "granule" system tells the position of each page and seek has to deal with that. Players get the whole duration of a file by reading the granule of the last page of the file.

1- When seeking within one track, headers must be added *once* at the beginning of the new position, so they must be gathered from the initial stream and then the seek must be done. Seek in librespot is done (AFAIK) by streaming the actual full data until the expected offset. I initially left some basic errors there as I misunderstood the way librespot chain tracks (see issue below) and there is no need to repeat the header when seeking, but the granule must always increment otherwise many players will terminate early. I don't think there is a perfect solution for granule handling when seeking (as one can seek forward and backward and this *not* a new instance of the decoder), so I've opted for a granule that always increases so that at least the player does not fall short. I'll continue thinking if there is a better but reasonably easy solution. 

2- When chaining tracks (without user pressing next), I initially thought that librespot was continuing with the same decoder instance, but it's not the case. A new decoder is created for every track, even when they flow "naturally" but so I have to keep a global context of granule if I want an incrementing value across tracks (per problem described above). I know when the last page/granule of a given stream is received as I have a last packet flag. Still, I cannot assume that at some point I won't receive a continuous data flow with multiple streams.

3- Last, when user changes tracks, librespot also creates a new decoder instance but of course in that case there is no last page so I need a way to continue this granule incrementation thingy.

Hope this helps and I'll continue thinking about options, mean while I'll push the updates I did

Johannesd3 commented on audio/src/lib.rs:
```diff
@@ -29,8 +30,65 @@ pub use fetch::{
+        }
+    }
+
+    pub fn empty(&self) -> bool {
+        match self {
+            AudioPacket::Samples(d) => d.len() == 0,
+            AudioPacket::Bytes(d) => d.len() == 0,
+        }
+    }
```
AFAIK, `is_empty` would be more usual in Rust. Vecs also have a function `is_empty` which you could use here. But that's bikeshedding.

philippe44 replied:
done

Johannesd3 replied:
Would you please also replace the `.len() == 0`? [Clippy would give a warning](https://rust-lang.github.io/rust-clippy/master/#len_zero) (and we have already enough of those warnings imo...)

sashahilton00 commented:
There's quite a bit of stuff to be tweaked here before it's ready to merge just looking over the discussion. Can I propose that given that this is not a breaking change we aim to get this in the `0.2.1` release? I'm conscious that the `librespot-tremor` issue for newer versions of rust is causing some problems for Spotifyd and would like to release a new version sooner rather than later.

philippe44 commented:
> There's quite a bit of stuff to be tweaked here before it's ready to merge just looking over the discussion. Can I propose we aim to get this in the `0.2.1` release? I'm conscious that the `librespot-tremor` issue for newer versions of rust is causing some problems for Spotifyd and would like to release a new version sooner rather than later.
>
You are the maintainer, I'll do as you recommend. My only ask is to not lose "momentum" on the discussion

philippe44 commented:
I've addressed all the comments I could but I'm not sure why I have this conflict on player.rs. Let me know if you want to restart a fresh PR or stash things together

Johannesd3 commented:
> There's quite a bit of stuff to be tweaked here before it's ready to merge just looking over the discussion. Can I propose we aim to get this in the `0.2.1` release? I'm conscious that the `librespot-tremor` issue for newer versions of rust is causing some problems for Spotifyd and would like to release a new version sooner rather than later.

@sashahilton00 Maybe apply/cherry-pick the tremor fix onto the current 0.1.3 and make a 0.1.4 of it?

philippe44 commented:
I've finally cleaned up (I hope for the best) the different commits and rebased to the latest dev to make differences easier to follow

philippe44 commented on playback/src/player.rs:
```diff
@@ -723,7 +723,19 @@ impl PlayerTrackLoader {
+                Err(e) => Err(AudioError::VorbisError(e)),
+            }
+        };
+
```
I'm not sure this is the best option but my Rust "rookieness" really hurts here

Johannesd3 replied:
I think it's fine. You _could_ use `Result::map` and  `map_err` instead of `match`, but I don't think it's a big win.

philippe44 replied:
thx!

Johannesd3 commented on audio/src/passthrough_decoder.rs:
```diff
@@ -0,0 +1,210 @@
+            }
+
+            // update global granule position
+            ABSGP_PAGE.fetch_add(absgp_page - self.absgp_page, Ordering::Relaxed);
```
Use the return value instead of loading it again below. Atomic operations are expensive.

philippe44 replied:
Done

Johannesd3 commented on audio/src/passthrough_decoder.rs:
```diff
@@ -0,0 +1,210 @@
+            let absgp_page = pck.absgp_page();
+
+            // set granule if unknown
+            if self.absgp_page == u64::MAX {
+                self.absgp_page = absgp_page;
+            }
```
Feels a bit hacky. Maybe `Option<u64>` would be clearer? And maybe rename the field to `last_absgp_page`?

philippe44 replied:
Done but what do you expect from a C developer :-) ?

philippe44 commented on audio/src/passthrough_decoder.rs:
```diff
@@ -0,0 +1,210 @@
+    return Ok(absgp_page);
+}
+
+static ABSGP_PAGE: AtomicU64 = AtomicU64::new(0);
```
Done

Johannesd3 commented on audio/src/passthrough_decoder.rs:
```diff
@@ -131,29 +133,28 @@ impl<R: Read + Seek> AudioDecoder for PassthroughDecoder<R> {
-            if self.absgp_page == u64::MAX {
-                self.absgp_page = absgp_page;
+            if self.lastgp_page.is_none() {
+                self.lastgp_page = Some(pckgp_page);
```
You could use something like 
```rust
let lastgp_page = *self.lastgp_page.get_or_insert(pckgp_page);
```

to avoid the `unwrap`s below.

philippe44 replied:
done

Johannesd3 commented:
I have no objections anymore. While the static granule seems not ideal, I don't know a better solution either.

philippe44 commented:
> I have no objections anymore. While the static granule seems not ideal, I don't know a better solution either.
>
Thank you for your comments - Rust is a steep learning curve :-)

Johannesd3 commented:
I don't know much about ogg, so this idea might be stupid: Create some kind of OggSink and keep track of the granule in there. The sink would get the vorbis data along with the granule differences. Would it work? However, this should not block this PR.

philippe44 commented:
So I think I've made all the requested changes, including in all backends + error management. I hope this is a better "Rust-compliant".
I've also made modification in the granule management because I was wrong, the granule does not need to be monotonic across chained streams, it only needs to be within the same stream (when seeking). Still, it was not working before because I did not realize that spotify does not set the stream serial number and in that case, the Ogg stream seems to be a single one and there, a monotonic granule was required. I've just added a fake stream serial if it is set to 0

Johannesd3 commented:
It's great that the problem with the global granule is solved, thanks!

Maybe you could run `cargo clippy` and fix the warnings that are related to this PR?

Johannesd3 commented on audio/src/passthrough_decoder.rs:
```diff
@@ -0,0 +1,195 @@
+    }
+}
+
+#[derive(Debug)]
+pub enum PassthroughError {
+    NotVorbisError,
+    OggError(OggReadError),
+    IOError(io::Error),
+}
```
I think the `IOError` case will never occur, but I won't guarantee it: This variant is used only if `PacketWriter<Vec<u8>>::write_packet` fails. Looking at the source code, it seems as if this function only passes through the errors of the underlying `Write` object. But writing to a `Vec` never fails, so you could as well `unwrap` the `Result`s of `write_packet`. (Is there anyone who could confirm this?)

If this variant is removed, you could also remove the `NotVorbisError` variant and replace it by `OggReadError::InvalidData`. So there wouldn't be a need for an enum anymore, you could simply use a "newtype" struct as in the vorbis decoders.

philippe44 replied:
yes, more simplification :-)

Johannesd3 commented on audio/src/lib.rs:
```diff
@@ -29,8 +30,70 @@ pub use fetch::{
+pub use libvorbis_decoder::{VorbisDecoder, VorbisError};
+pub use passthrough_decoder::{PassthroughDecoder, PassthroughError};
+
+#[derive(Debug)]
+pub enum AudioError {
+    PassthroughError(PassthroughError),
+    VorbisError(VorbisError),
+}
```
Is it the right way to make `AudioDecoder` a trait, but `AudioError` an enum?

philippe44 replied:
done then

Johannesd3 reviewed (commented):
I added a few comments regarding error handling. 

@sashahilton00 @ashthespy Do you think this would be an appropriate place to start using `thiserror`? It would spare us a lot of boilerplate here.

philippe44 commented on audio/src/passthrough_decoder.rs:
```diff
@@ -0,0 +1,195 @@
+
+impl fmt::Display for PassthroughError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        fmt::Display::fmt(&self, f)
```
This is the same as the others (lewton & libvorbis)

Johannesd3 replied:
No, there it's `self.0`, so it is forwarded to the inner type. Here, it's forwarded to itself.

philippe44 replied:
argh....

philippe44 commented:
Let me know if there is something else expected from me

sashahilton00 commented:
Just released `0.1.5` with the tremor fix, so there isn't pressure on this PR anymore. Will wait to hear from anyone that has any further thoughts/suggestions, otherwise this looks pretty much ready to go @philippe44

sashahilton00 commented:
@philippe44 would you mind squashing the commits?

Johannesd3 commented:
May I point out that the commit has a not very meaningful name?

sashahilton00 commented:
If you could rename the commit to something more meaningful then that would be good. After that I'll merge it.

Johannesd3 commented:
@ashthespy (or @sashahilton00) After this is merged, could you try merging `dev` into `tokio_migration`, if you can spare the time? The more they diverge, the harder it will be to merge eventually.

sashahilton00 commented:
Merged. Please update the wiki accordingly as requested above.

philippe44 commented:
> Merged. Please update the wiki accordingly as requested above.
>
Done
